### ISSUE TITLE
Get Customer Devices

### ISSUE THREAD

**Issue by xTwigx** [thingsboard/thingsboard#11506]
I have an HTML widget for a side menu for my dashboard to navigate between states.

in the side menu i have a submenu for the customer to navigate to different reports. 
![image](https://github.com/user-attachments/assets/0eface65-3351-4f7c-9a4b-9c1d0a5c9447)

Now this submenu is shown via a custom action in the widget 
`
    // Find the submenu and arrow elements
    var reportsSubmenu = widgetContext.$container.find('#reportsSubmenu');
    var reportsArrow = widgetContext.$container.find('#reportsArrow');

    // Toggle submenu visibility
    if (reportsSubmenu.is(':visible')) {
        reportsSubmenu.hide();
        reportsArrow.removeClass('rotate');
    } else {
        reportsSubmenu.show();
        reportsArrow.addClass('rotate');
    }
`
I need This to be slightly change so the customer any sees reports submenus depending of the type of devices he has. Ex if he doesn't have a water salinity device he doesn't see the Water Salinity Reports Sub menu.

I was wondering If i can get any help on how to do this. Or how to get the devices that the user has and check the type ?

 * ThingsBoard: 3.7.0 PE

**Comment by xTwigx** [comment]
Could there also be simple way to get the customerName in the widget and maybe i could simply put it into a switch/if statement to see what to display?

**Comment by devaskim** [comment]
It is possible to implement either with HTML/Markdown widget where you can specify JS function to build HTML part of the widget, or with custom widget where you are not limited by nothing

**Comment by xTwigx** [comment]
Yes i know that menu is made with a HTML/Markdown But the issue is i want to modify/add to the menu depending on the customerDevices (types). My main issue/question is how do i get the customerDevices?

**Comment by devaskim** [comment]
The solution depends on who will open dashboard:
1. Customer User -> use Entity Alias of `Device` Entity Type in menu widget
2. Tenant Admin -> create relation between customer and his devices, use Entity Alias of `Relations Query` Type in menu widget

**Comment by xTwigx** [comment]
Im not a 100% we understand each other perfectly.

1) Im building a website looking dashboard. 
![image](https://github.com/user-attachments/assets/176b3be7-ff33-4b7f-9488-6824b0ee8202)

2) this dashboard will be given access to multiple customer for them to see devices on.
3) There are multiple types of devices ( currently 2 energy monitoring, water salinity)
4) The customer don't all have a device of each some customer can have only water Salinity and another customer can have energy Monitoring devices.
5) In the Reports Submenu I have 2 report types (water and energy) they each go to a specific dashboard state that has different widget/charts 
6) Since the customer wont have a type of device (lets say he only has a Energy Device) i don't want him to be able to have the water salinity reports submenu.

Hope this made it a bit more clear, appreciate you tying to help

**Comment by devaskim** [comment]
I think I understood question `Get Customer Devices` and provided possible solutions, but you need the answer for `Get Types of Customer Devices`, right?

**Comment by xTwigx** [comment]
Yes, for example i get all of the devices that are owned by the customer who logged in and have they saved in an array or something and i can go through them and check for the types the devices are.

The issue is i need the devices in a "code format"

**Comment by devaskim** [comment]
Unfortunately, there is no easy way to implement such functionality. I know 2 ways:
1. Based on Rule Chains: catch `Entity Assigned` and `Entity Unassigned` events, check customer ownership of this entity and entity type. Then update, for example, customer attribute with the array of unique device types of this customer. 

2. Based on Widgets (pure performance): each time the (widget) loads, iterate through all customer devices and collect array of unique device types.
======
Finally having the array of unique device type, you can use in menu widget to form report types submenu.

**Comment by xTwigx** [comment]
I see i will try to do something similar thank you for the help. I am going to leave the issue open just incase someone has another way :)

**Comment by xTwigx** [comment]
I found this in the action . var customerTitle = widgetContext.customerService.getCustomer;

but when i use it to try and get the customer i get this: getCustomer(l,c){return this.http.get(`/api/customer/${l}`,(0,t.T7)(c))}

**Comment by devaskim** [comment]
Hm, it is programming basics. The `getCustomer` is JavaScript method with 1 obligatory parameter and 1 option parameter. So to call this method, you should do at least `widgetContext.customerService.getCustomer(param1)`

https://github.com/thingsboard/thingsboard/blob/master/ui-ngx/src/app/core/http/customer.service.ts#L39

**Comment by xTwigx** [comment]
I see it excepts the customerId as a parameter but i dont understand how i should be getting the customer id or the title

**Comment by devaskim** [comment]
As alternative to programming way of getting customer title, you can create Entity Alias of `Current Customer` Type and use it in your widget's data source

**Comment by xTwigx** [comment]
You you please show me how do this because i cant seem to get it.

My hierarchy is like this so you know
Customer -> subcustomer -> subcustomer

If there is a way to make an alias to just get the customer you are in ONLY that would be perfect

**Comment by devaskim** [comment]
https://github.com/user-attachments/assets/07b489a6-c0d8-4a90-9466-689b8ad6ceae

**Comment by xTwigx** [comment]
I see Thanks Very much i managed to do something similar to what i wanted .

Just so other know i used that alias and took the name as a key. Then in the action im using that key to check with array i have that have the customers.

each array is then checked against the nameKey. If its true then it displays the menu i want

**Comment by devaskim** [comment]
If it is OK for business, you could share your solution as widget json dump here. And I would add it to https://github.com/devaskim/awesome-thingsboard

**Comment by xTwigx** [comment]
Here you go Hope this can help people. I have you understand its in no means perfect :)

[ConditionalReportMenu.json](https://github.com/user-attachments/files/16797936/ConditionalReportMenu.json)

**Comment by xTwigx** [comment]
Not A problem hope it can help someone else. And once again Thank you very much for your immense help.

**Comment by mykhailokornieiev** [comment]
I closed this ticket due to inactivity. If this question is still relevant to you, feel free to open a new ticket with full description of your request
